### PR TITLE
[X86] Add support for indirect branch tracking in jump tables

### PR DESCRIPTION
**Pull request by nmosier**
This patch adds support for protecting jump tables with indirect branch tracking (IBT). By default, indirect jump table branches are given a 'notrack' prefix and thus not protected with IBT. This default behavior is suitable for traditional threat models, assuming the compiler generates correct jump table code. However, for threat models encompassing speculative execution vulnerabilities (e.g., Spectre), such notrack'ed indirect branches potentially introduce Spectre-v2-style vulnerabilites by allowing them to mis-speculatively jump to arbitrary target addresses, not just ENDBRANCH instructions.

To enable indirect branch tracking for jump tables, this patch allows the `cf-protection-branch` module flag's value to indicate the level of indirect branch protection required. If `cf-protection-branch == 0` or `cf-protection-branch` is missing entirely, then branch protections are applied. If `cf-protection-branch == 1`, then branch protections are applied for all indirect branches except for those that cannot under any circumstances non-speculatively jump to the wrong target (namely, indirect jump table branches). If `cf-protection-branch >= 2`, then branch protections are applied to all indirect branches, including indirect jump table branches.

To summarize the new interpretation of the `cf-protection-branch` module flag under this patch:
* `cf-protection-branch == 1` (currently emitted by clang when passed flag `-fcf-protection=branch`) is suitable for hardening code against _non-speculative_ control-flow hijacks.
* `cf-protection-branch >= 2` is suitable for additionally hardening code against _speculative_ control-flow hijacks (e.g., Spectre v2).

### CODE REVIEW

File: llvm/lib/Target/X86/X86IndirectBranchTracking.cpp [hunk 1]
```diff
@@ -186,5 +197,15 @@ bool X86IndirectBranchTrackingPass::runOnMachineFunction(MachineFunction &MF) {
...
   }
+
+  // If strong CF protections are enabled (Level > 1), then add ENDBRs to all
+  // jump table BBs, since jump table branches will not have the 'notrack'
+  // prefix.
+  if (CFProtectionLevel > 1)
+    if (const MachineJumpTableInfo *JTI = MF.getJumpTableInfo())
+      for (const MachineJumpTableEntry &JTE : JTI->getJumpTables())
+        for (MachineBasicBlock *MBB : JTE.MBBs)
+          addENDBR(*MBB, MBB->begin());
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86IndirectBranchTracking.cpp:208]: `Changed |= addENDBR(*MBB, MBB->begin());`

File: llvm/lib/Target/X86/X86.h [hunk 2]
```diff
@@ -199,6 +199,7 @@ void initializeX86ReturnThunksPass(PassRegistry &);
 void initializeX86SpeculativeExecutionSideEffectSuppressionPass(PassRegistry &);
 void initializeX86SpeculativeLoadHardeningPassPass(PassRegistry &);
 void initializeX86TileConfigPass(PassRegistry &);
+void initializeX86IndirectBranchTrackingPassPass(PassRegistry &);
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86.h:202]: Is the change related to jump tables? Put it in a separate patch if not.

File: llvm/test/CodeGen/X86/indirect-branch-tracking-jt.mir [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# RUN: llc %s -o - -mtriple=x86_64-- -run-pass=x86-ibt | FileCheck %s
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/indirect-branch-tracking-jt.mir:1]: Check for both `cf-protection-branch` equals to 1 and 2? You can you sed to do so, e.g.,
```
# RUN: sed 's/LEVEL/1/g' %s | llc %s -o - -mtriple=x86_64-- -run-pass=x86-ibt | FileCheck %s --check-prefixes=L1
# RUN: sed 's/LEVEL/2/g' %s | llc %s -o - -mtriple=x86_64-- -run-pass=x86-ibt | FileCheck %s --check-prefixes=L2
...
  !0 = !{i32 1, !"cf-protection-branch", i32 LEVEL}
```

**Review by phoebewang — COMMENTED** [review]
Currently, Clang only set `cf-protection-branch` to 1, do you have follow up patch to update it?

**Comment by nmosier** [comment]
> Currently, Clang only set `cf-protection-branch` to 1, do you have follow up patch to update it?

Yes, I've prepared that as for a follow-up patch.

File: llvm/lib/Target/X86/X86.h (hunk 2 above)
Comment (nmosier) [reply llvm/lib/Target/X86/X86.h:202]: I added that so that it could be run via `-run-pass=x86-ibt`, as required in the test I added (indirect-branch-tracking-jt.mir). I can submit separate patch for it, but this patch will then be dependent on that one.

Comment (nmosier) [reply llvm/lib/Target/X86/X86.h]: Actually, I rewrote the test at the IR-level so running the x86-ibt pass via `-run-pass=x86-ibt` is no longer necessary. Thus, the above change (as well as `INITIALIZE_PASS(X86IndirectBranchTracking, ...)`), etc.) is no longer necessary, so I've reverted those.

File: llvm/test/CodeGen/X86/indirect-branch-tracking-jt.mir (hunk 3 above)
Comment (nmosier) [reply llvm/test/CodeGen/X86/indirect-branch-tracking-jt.mir]: The new test now checks for cf-protection-branch =  {0, 1, 2}.

File: llvm/lib/Target/X86/X86IndirectBranchTracking.cpp (hunk 1 above)
Comment (nmosier) [reply llvm/lib/Target/X86/X86IndirectBranchTracking.cpp]: Thanks for the catch! I addressed that in the latest commit.

**Review by phoebewang — APPROVED** [review]
LGTM, thanks!

**Comment by MaskRay** [comment]
I hope that we don't add a new mode. Armv8-M and AArch64's Branch Target Identification just provide one mode where a switch table's destinations are marked with BTI. Can x86 IBT do the same?

**Comment by nmosier** [comment]
Yeah, we could do the same for X86. But I think having two modes, one per threat model, is desireable, since adding ENDBRANCH instructions to each jump table block can be very expensive (e.g., it can double the jump table code size if each block originally only has 1–2 instructions). That way, code that doesn't care about Spectre et al. can forgo jump table protections without suffering any (modest) performance impact.

I'm open to either option, though. The project I'm working on right now requires IBT protection for all indirect branches, anyway, so having a single mode that unconditionally protects indirect jump table branches works just as well.

**Review by MaskRay — CHANGES_REQUESTED** [review]
I think this needs more discussions. Request change so that this does not land before the discussion concludes. An extra option needs coordination with GCC.

https://github.com/llvm/llvm-project/issues/54247
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104816

**Comment by MaskRay** [comment]
It looks like GCC provides `-mcet-switch` to remove NOTRACK in favor or ENDBR in every case handler. It defaults to off, however, as I understand the overwhelming opinion on https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104816 and AArch64 BTI's preference, `-mcet-switch` should default to on and there is actually no point providing the option. In Clang, perhaps we should just always use ENDBR to simplify LTO handling.
